app-common/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+use clap::Args;
+use rustls::{ClientConfig, ServerConfig, SupportedProtocolVersion, pki_types::{CertificateDer, CertificateRevocationListDer, PrivateKeyDer, pem::PemObject}, server::WebPkiClientVerifier};
+use tracing::{debug, info};
+use std::{path::PathBuf, sync::Arc};
+use snafu::{Snafu, ResultExt};
+
+#[derive(Snafu, Debug)]
+pub enum TlsError {
+    #[snafu(display("IO error"))] 
+    Io { source: std::io::Error },
+    #[snafu(display("PEM parse error in path: {}", path.as_ref().map(|p| p.display().to_string()).unwrap_or("unknown".into())))]
+    PemParse { 
+        source: rustls::pki_types::pem::Error,
+        path: Option<PathBuf>,
+     },
+    #[snafu(display("Rustls error"))]
+    Rustls { source: rustls::Error },
+
+    #[snafu(display("Certificate verifier error"))]
+    CertificateVerifier { source: rustls::client::VerifierBuilderError},
+
+    #[snafu(display("Config error"))]
+    Config { message: String },
+}
+
+
+#[derive(Args, Debug)]
+pub struct TLSOptions {
+    /// Enables mTLS (TLS for DICOM connections)
+    #[arg(long = "tls", default_value = "false")]
+    pub enabled: bool,
+
+    /// Crypto provider to use, see documentation (https://docs.rs/rustls/latest/rustls/index.html) for details
+    #[arg(long, value_enum, default_value_t = CryptoProvider::AwsLC, value_name = "provider")]
+    pub crypto_provider: CryptoProvider,
+
+    /// List of cipher suites to use. If not specified, the default cipher suites for the selected crypto provider will be used.
+    #[arg(long, value_name = "cipher1,...")]
+    pub cipher_suites: Option<Vec<String>>,
+
+    /// TLS protocol versions to enable
+    #[arg(long, value_enum, value_name = "version,...", default_values_t = vec![TLSProtocolVersion::TLS1_2, TLSProtocolVersion::TLS1_3])]
+    pub protocol_versions: Vec<TLSProtocolVersion>,
+
+    /// Path to private key file in PEM format
+    #[arg(long, value_name = "/path/to/key.pem,...")]
+    pub key: Option<PathBuf>,
+
+    /// Path to certificate file in PEM format
+    #[arg(long, value_name = "/path/to/cert.pem,...")]
+    pub cert: Option<PathBuf>,
+
+    /// Path to additional CA certificates (comma separated) in PEM format to add to the root store
+    #[arg(long, value_name = "/path/to/cert.pem,...")]
+    pub add_certs: Option<Vec<PathBuf>>,
+
+    /// Add Certificate Revocation Lists (CRLs) to the server's certificate verifier
+    #[arg(long, value_name = "/path/to/crl.pem,...")]
+    pub add_crls: Option<Vec<PathBuf>>,
+
+    #[arg(long, action = clap::ArgAction::SetFalse)]
+    /// Load certitificates from the system root store
+    pub system_roots: bool,
+
+    /// How to handle peer certificates
+    #[arg(long, value_enum, value_name = "opt", default_value_t = PeerCertOption::Require)]
+    pub peer_cert: PeerCertOption,
+
+    /// Allow unauthenticated clients (only valid for server)
```

Comment:
It's a bit of a bummer that we have options which only make sense to connection acceptors. Would there be a way to hide it in `dicom-storescu`? I suppose this would require different types?

Reply:
I was thinking of having a `TlsCommon` struct and then one for Acceptors vs. Initiators, but I only found one field that would be in either of those which is this one. I can still do that though.